### PR TITLE
build: use `uv` and `ruff`

### PR DESCRIPTION
**Pull request by slevang**
So this would be replacing `poetry` with yet another new shiny tool, but I would argue the benefits are worth it. `uv` is a drop-in replacement for `pip`, and `ruff` is their linter/formatter tool. Both are incredibly fast (written in rust), have a lot of convenient options, and are seeing rapid adoption across the python ecosystem.

One really nice feature of `uv` is the `--resolution` arg which I've used here to run tests on a "minimal" version, using lowest supported python version of 3.10 and lowest specified versions of all our primary dependencies. Then the 3.11 and 3.12 jobs run with `--resolution=highest` which will always have us testing against new versions.

The substantive changes of this PR are the `pyproject.toml` and the GH Actions. Everything else is only the result of running ruff's linter (almost all of them auto-fixed). Mostly unused imports, trimming whitespace, and a few other random things like truthy/falsy checks and unused f-strings.

I did mess with the docs a little to get that running, and rebuilt all the sphinx. Hopefully didn't screw anything up there.

Only thing I haven't tested is the `build_command` and uploading of wheels to pypi, but it should work fine.

### CODE REVIEW

**Comment by nicrie** [comment]
Hey @slevang that's awesome ! I heard about these things recently coming out from Astral and would love to give it a try and see how they work. Expecting my first child literally today :)  - so I won't have time to review anytime soon. I'm OK giving it a try 👍

**Comment by slevang** [comment]
Amazing, congrats! Hope all goes smoothly, and enjoy those early days :smiling_face_with_three_hearts:

File: docs/environment.yml [hunk 1]
```diff
@@ -3,27 +3,9 @@ channels:
...
-  - sphinx-design
-  - sphinx-copybutton
-  - nbsphinx
-  - pydata-sphinx-theme
-  - sphinx
-  - nbconvert
+  - pandoc
+  - pip
   - pip:
-    - xeofs
```
Comment (nicrie) [inline docs/environment.yml:29]: Nice! This is so much cleaner! =)

File: docs/environment.yml [hunk 2]
```diff
@@ -3,27 +3,9 @@ channels:
...
+    - -e ../.[docs]
```
Comment (nicrie) [inline docs/environment.yml:11]: Nice! This is so much cleaner!

**Review by nicrie — APPROVED** [review]
That's great - looking forward to see it in action! Thank you @slevang !

**Review by nicrie — CHANGES_REQUESTED** [review]
One additional thought: we probably want to update the contributing guide, specifically the section about [how to create a local env](https://github.com/xarray-contrib/xeofs/blob/main/docs/overview_3_contributing.rst#2-set-up-your-development-environment). Instead of poetry, would you prefer `pip install -e .` and `pip install -e .[docs]` or rather `uv`?

**Comment by slevang** [comment]
Good point. I went with basic `pip` to keep things simple. I also had to add those two system dependencies to the conda env creation to get the docs dependencies to install properly.

**Review by nicrie — APPROVED** [review]

**Pull request merged**